### PR TITLE
CA-394109: Reduce number of alerts

### PR DESCRIPTION
**Pull request by Vincent-lau**
relatively large change, best reviewed by commit

### CODE REVIEW

File: ocaml/idl/datamodel_cluster.ml [hunk 1]
```diff
@@ -169,6 +169,16 @@ let pool_resync =
     ~params:[(Ref _cluster, "self", "The cluster to resync")]
     ~lifecycle ~allowed_roles:_R_POOL_OP ~errs:[] ()
 
+let cstack_sync =
+  call ~name:"cstack_sync"
+    ~doc:
+      "Sync xapi db with the cluster stack synchronously, generate alerts as \
```
Comment (lindig) [inline ocaml/idl/datamodel_cluster.ml:175]: I would have said ".. and generate ..".

File: ocaml/idl/datamodel_cluster.ml [hunk 2]
```diff
@@ -169,6 +169,16 @@ let pool_resync =
...
+       needed. Only happens on the coordinator as this is where the cluster \
+       watcher perform updates."
```
Comment (lindig) [inline ocaml/idl/datamodel_cluster.ml:177]: "performs" (singular)?

File: ocaml/xapi/xapi_clustering.ml [hunk 3]
```diff
@@ -426,133 +426,168 @@ let compute_corosync_max_host_failures ~__context =
...
-                      __FUNCTION__ ipstr ;
-                    None
-                | Some ch ->
-                    Some ch
-              )
-              nodel
+module Watcher = struct
+  let on_corosync_update ~__context ~cluster updates =
+    debug
+      "%s: Received %d updates from corosync_notifyd , run diagnostics to get \
```
Comment (lindig) [inline ocaml/xapi/xapi_clustering.ml:432]: space before comma: remove

File: ocaml/xapi/xapi_clustering.ml [hunk 4]
```diff
@@ -426,133 +426,168 @@ let compute_corosync_max_host_failures ~__context =
...
+          let live_hosts =
+            Db.Cluster_host.get_refs_where ~__context
+              ~expr:(Eq (Field "live", Literal "true"))
           in
-          let missing_hosts =
-            List.filter
-              (fun h -> not (List.mem h quorum_hosts))
-              all_cluster_hosts
+          let dead_hosts =
+            List.filter (fun h -> not (List.mem h live_hosts)) all_cluster_hosts
```
Comment (lindig) [inline ocaml/xapi/xapi_clustering.ml:449]: Should we use sets here? This is O(n^2).

File: ocaml/xapi/xapi_clustering.ml [hunk 5]
```diff
@@ -426,133 +426,168 @@ let compute_corosync_max_host_failures ~__context =
...
+                    ~value:current_time
+                )
+                all_cluster_hosts
+          | Some nodel ->
+              (* nodel contains the current members of the cluster, according to corosync *)
+              let quorum_hosts =
+                List.filter_map
+                  (fun {addr; _} ->
+                    let ipstr = ipstr_of_address addr in
+                    match List.assoc_opt ipstr ip_ch with
```
Comment (lindig) [inline ocaml/xapi/xapi_clustering.ml:479]: Again O(n^2); is this efficient enough?

**Review by lindig — APPROVED** [review]

File: ocaml/xapi/xapi_clustering.ml (hunk 4 above)
Comment (Vincent-lau) [reply ocaml/xapi/xapi_clustering.ml:449]: the maximum number of hosts we support is 16, so maximum complexity is 256, should be fine, what do you think?

**Comment by BengangY** [comment]
If you use VSCode, you can configure setting.json to auto-format when saving the code like this:
```
    "[ocaml]": {
        "editor.formatOnSave": true
    },
```

File: ocaml/xapi/xapi_clustering.ml [hunk 6]
```diff
@@ -522,11 +522,16 @@ module Watcher = struct
 
   let cluster_change_watcher : bool Atomic.t = Atomic.make false
 
+  (* this is the time it takes for the update request to time out. It is ok to set
+     it to a relatively long value since the call will return immediately if there
+     is an update *)
+  let cluster_change_interval = Mtime.Span.min
```
Comment (psafont) [inline ocaml/xapi/xapi_clustering.ml:541]: note, min is for minute, not minimum

Comment (Vincent-lau) [reply ocaml/xapi/xapi_clustering.ml:541]: 👍 which I think I have used in the right way

**Review by psafont — APPROVED** [review]
This change replaces the direct call to alerts from normal operation to trigger a stack synchronization. This should be caught by the watcher and generate the alert once, instead of twice like before

**Comment by Vincent-lau** [comment]
going to run a quick xrt test to make sure these changes are good, since it has been here fow a while

**Comment by Vincent-lau** [comment]
Added a feature flag, and tests look ok, merge this one now.

**Pull request merged**